### ISSUE TITLE
Overfit

### ISSUE THREAD

**Issue by jorahn** [jorahn/rookworld-trl#4]
@claude review the manual debug code and experiment results.
Come up with strategies to successfully overfit on a single batch. From batch size to algorithmic changes, anything goes

**Comment by jorahn** [comment]
Author: openai codex

Title: Manual Debug Overfit – True GA applied, quick validation, and next steps

Summary
- Switched manual_grpo_debug.py to true gradient accumulation (one optimizer step per batch); removed intra‑batch micro‑updates.
- Updated GA logging/CLI help and pyproject URLs. Ran 2‑step sanity checks (3 configs) to confirm behavior.
- Recommend 50‑step overfit with gens 24 vs 32; keep stochastic training and small entropy.

Changes
- True GA: accumulate per‑sample loss scaled by 1/(batch_size·gens), then a single clip/step/zero per batch per GRPO step. Removed optimizer.step() at microbatch boundaries.
- Logging: Banner now states “GA: true accumulation; one optimizer step per batch; effective_batch=…”. If `--ga != 1`, print a note that it’s ignored for stepping. Removed micro‑update prints.
- CLI help: `--grad_accum_steps/--ga` clarified as reserved (ignored for stepping under true GA).
- URLs: pyproject.toml now points to https://github.com/jorahn/rookworld-trl.

Quick Validation (2 steps, P‑only fixed batch)
- A: bs=2, gens=4, ga=1 → 0.0786 → 0.0786 (Δ +0.0000). One optimizer update per step confirmed.
- B: bs=2, gens=4, ga=4 → identical behavior/outcome (ga ignored for stepping under true GA).
- C: bs=2, gens=8, ga=1 → 0.0786 → 0.0932 (Δ +0.0146). Slight positive movement with higher gens even over 2 steps.

Why this matters
- Multiple micro‑steps per batch bias updates (stateful Adam, repeated clipping, model drift mid‑batch). For single‑batch overfit, true GA yields a coherent update on the frozen batch objective and can reduce oscillation.

Recommendations (Overfit One Batch)
- Keep stochastic training generation (do_sample=True) and a small entropy term (0.003–0.005) to preserve within‑group variance for GRPO advantages. Avoid deterministic gen and entropy=0.
- Run two 50‑step configs on the same P‑only fixed batch:
  1) Baseline: bs=8, gens=24, warmup=20, entropy=0.005, beta_after=0.005, lr=2e‑6
  2) High‑gens: bs=8, gens=32, warmup=10, entropy=0.003, beta_after=0.0025–0.005, lr=2e‑6
- Monitor: 20‑step moving average (greedy reward) slope, KL% (target ~20–60%), grad norms stability, and advantage variance. If slope is flat and stability holds, bump lr→3e‑6.

Suggested Commands
```
# Baseline (50 steps)
uv run python manual_grpo_debug.py \
  --overfit_single_batch --steps 50 --batch_size 8 --gens 24 \
  --ga 1 --beta_warmup_steps 20 --entropy_coef 0.005

# High-gens (50 steps)
uv run python manual_grpo_debug.py \
  --overfit_single_batch --steps 50 --batch_size 8 --gens 32 \
  --ga 1 --beta_warmup_steps 10 --entropy_coef 0.003
```

Notes
- `--ga` is currently ignored for stepping (kept for future chunking); true GA always performs a single optimizer step per batch.
- EXPERIMENTS_2025‑09‑03_LOG.md updated with the change and the quick validation results.
